src/concurrency/vector_clock.rs

```diff
@@ -52,25 +52,33 @@ const SMALL_VECTOR: usize = 4;
 /// so that diagnostics can report what code was responsible for an operation.
 #[derive(Clone, Copy, Debug)]
 pub struct VTimestamp {
-    // The lowest bit indicates read type, the rest is the time.
+    /// The lowest bit indicates read type, the rest is the time.
+    /// `1` indicates a retag read, `0` a regular read.
     time_and_read_type: u32,
     pub span: Span,
 }
 
 impl VTimestamp {
-    pub const ZERO: VTimestamp = VTimestamp { time_and_read_type: 0, span: DUMMY_SP };
+    pub const ZERO: VTimestamp = VTimestamp::new(0, NaReadType::Read, DUMMY_SP);
+
+    const fn new(time: u32, read_type: NaReadType, span: Span) -> Self {
+        let read_type = match read_type {
+            NaReadType::Read => 0,
+            NaReadType::Retag => 1,
+        };
+        Self {
+            time_and_read_type: read_type | time.checked_mul(2).expect("Vector clock overflow"),
+            span,
+        }
```

Comment:
Is there any way to avoid duplicating the logic with the `set` functions?

Reply:
I could call `new` in them. It might result in code for dealing with the span though.